plansys2_executor/src/plansys2_executor/ActionExecutor.cpp

```diff
@@ -207,6 +210,19 @@ ActionExecutor::tick(const rclcpp::Time & now)
   return get_status();
 }
 
+void
+ActionExecutor::cancel()
```

Comment:
This message implements the protocol described [here](https://github.com/IntelligentRoboticsLabs/ros2_planning_system/issues/69#issuecomment-761537077). This msg goes to the action performer, so it can't be filled.

Reply:
I think I understand. Actions are handled from 2 sides. The 1st is the performer/client side, which is instantiated by sub-classing the ActionExecutorClient. The 2nd is the executor side, which is an instantiation of ActionExecutor. The two sides communicate via the /action_hub message.

So you're saying that it would not be appropriate to fill in the completion, status, and success fields on the executor side, since those fields are reserved for the performer side, i.e., the performer should tell the executor about its progress not the other way around. Is that correct?

However, would the cancel() function be called from the executor side? In that case, would it be appropriate to inform the performer that the action was canceled on the executor side? One way that I was thinking about using this is to allow the ActionExecutor to cancel the execution if the actual duration of an action exceeded the predicted duration (as specified by the plan) by some percentage. In this scenario, I think cancel() would be invoked from the ActionExecutor.

Reply:
> So you're saying that it would not be appropriate to fill in the completion, status, and success fields on the executor side, since those fields are reserved for the performer side, i.e., the performer should tell the executor about its progress not the other way around. Is that correct?

Exactly

> However, would the cancel() function be called from the executor side? In that case, would it be appropriate to inform the performer that the action was canceled on the executor side? 

We do it. An ActionExecution.msg is sent from the Executor side, setting the type field to CANCEL. These are the possible values for type (from ActionExecution.msg):

`
int16 REQUEST=1
int16 RESPONSE=2
int16 CONFIRM=3
int16 REJECT=4
int16 FEEDBACK=5
int16 FINISH=6
int16 CANCEL=7
`

> One way that I was thinking about using this is to allow the ActionExecutor to cancel the execution if the actual duration of an action exceeded the predicted duration (as specified by the plan) by some percentage. In this scenario, I think cancel() would be invoked from the ActionExecutor.

This is interesting. But I am not sure if using the duration specified in the domain file is adequate. The duration of some actions is impossible to really predict (navigation, for example). Maybe we could add an optional parameter in the ActionExecutionClient to set a timeout...